mshv-ioctls/src/ioctls/system.rs

```diff
@@ -0,0 +1,201 @@
+use libc::{open, O_CLOEXEC, O_NONBLOCK};
+use mshv_bindings::*;
+use std::fs::File;
+use std::os::raw::c_char;
+use std::os::unix::io::{FromRawFd, RawFd};
+use vmm_sys_util::errno;
+use vmm_sys_util::ioctl::ioctl_with_ref;
+
+/// Wrapper over MSHV system ioctls.
+pub struct Mshv {
+    hv: File,
+}
+
+impl Mshv {
+    ///
+    /// Opens `/dev/mshv` and returns a `Mshv` object on success.
+    ///
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> Result<Self> {
+        // Open `/dev/mshv` using `O_CLOEXEC` flag.
+        let fd = Self::open_with_cloexec(true)?;
+        // Safe because we verify that ret is valid and we own the fd.
+        let ret = unsafe { Self::new_with_fd_number(fd) };
+        let version_result = ret.request_version();
+        match version_result {
+            Ok(_) => Ok(ret),
+            Err(err) => Err(err),
+        }
+    }
+    /// Creates a new Mshv object assuming `fd` represents an existing open file descriptor
+    /// associated with `/dev/mshv`.
+    ///
+    pub unsafe fn new_with_fd_number(fd: RawFd) -> Self {
+        Mshv {
+            hv: File::from_raw_fd(fd),
+        }
+    }
+
+    ///
+    /// Opens `/dev/mshv` and returns the fd number on success.
+    ///
+    pub fn open_with_cloexec(close_on_exec: bool) -> Result<RawFd> {
+        let open_flags = O_NONBLOCK | if close_on_exec { O_CLOEXEC } else { 0 };
+        // Safe because we give a constant nul-terminated string and verify the result.
+        let ret = unsafe { open("/dev/mshv\0".as_ptr() as *const c_char, open_flags) };
+        if ret < 0 {
+            Err(errno::Error::last())
+        } else {
+            Ok(ret)
+        }
+    }
+    ///
+    /// Creates a VM fd using the MSHV fd.
+    ///
+    pub fn create_vm(&self) -> Result<VmFd> {
+        // Safe because we know `self.hv` is a real MSHV fd as this module is the only one that
+        // creates mshv objects.
+        let creation_flags: u64 = HV_PARTITION_CREATION_FLAG_LAPIC_ENABLED as u64
+            | HV_PARTITION_CREATION_FLAG_EXO_PARTITION as u64;
+        let pr = mshv_create_partition {
+            partition_creation_properties: hv_partition_creation_properties {
+                disabled_processor_features: hv_partition_processor_features { as_uint64: [0; 2] },
+                disabled_processor_xsave_features: hv_partition_processor_xsave_features {
+                    as_uint64: 0 as __u64,
+                },
+            },
+            flags: creation_flags,
+        };
+        let ret = unsafe { ioctl_with_ref(&self.hv, MSHV_CREATE_PARTITION(), &pr) };
+        if ret >= 0 {
+            // Safe because we verify the value of ret and we are the owners of the fd.
+            let vm_file = unsafe { File::from_raw_fd(ret) };
+            Ok(new_vmfd(vm_file))
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+    ///
+    /// Validate the MSHV version
+    ///
+    pub fn request_version(&self) -> Result<()> {
```

Comment:
Shouldn't we allow to pass a version as an argument to this function? I know for now we only want to support v0, but in the future you may want to allow for supporting different binding versions.